sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala

```diff
@@ -48,7 +48,7 @@ object RemoveRedundantProjects extends Rule[SparkPlan] {
   private def removeProject(plan: SparkPlan, requireOrdering: Boolean): SparkPlan = {
     plan match {
       case p @ ProjectExec(_, child) =>
-        if (isRedundant(p, child, requireOrdering)) {
+        if (isRedundant(p, child, requireOrdering) && canRemove(p, child)) {
           val newPlan = removeProject(child, requireOrdering)
           newPlan.setLogicalLink(child.logicalLink.get)
```

Comment:
how about use

`newPlan.setLogicalLink(child.logicalLink.get)` -> `newPlan.setLogicalLink(p.logicalLink.get)` ?

Reply:
This looks risky. It makes it hard to reason about the mapping between physical and logical plans.

Reply:
I think we should even remove this line